### PR TITLE
Implementation of `NoEvents` and `CEP47` modality 

### PR DESCRIPTION
**Pull request by darthsiroftardis**
This is a retargetting of the PR #146 to the correct feature branch

### CODE REVIEW

File: contract/src/error.rs [hunk 1]
```diff
@@ -142,6 +142,19 @@ pub enum NFTCoreError {
...
+    InvalidTokenEventOrder = 143,
+    FailedToRecordMintEvent = 144,
+    FailedToRecordBurnedEvent = 145,
+    FailedToRecordApproveEvent = 146,
+    FailedToRecordTransferEvent = 147,
+    FailedToRecordMetadataUpdateEvent = 148,
+    InvalidStartingEventId = 149,
+    InvalidLastEventId = 150,
+    MissingEventsMode = 151,
+    InvalidEventsMode = 152,
```
Comment (gRoussac) [inline contract/src/error.rs]: Error excepted 
```
MissingTokenEventId 
InvalidTokenEventId
InvalidTokenEventOrder
FailedToRecordMintEvent
FailedToRecordApproveEvent
FailedToRecordTransferEvent
InvalidStartingEventId
InvalidLastEventId
```

do not seem to be used

```suggestion
    InvalidTokenEvent = 140,
    MissingEventsMode = 141,
    InvalidEventsMode = 142,
```

File: README.md [hunk 2]
```diff
@@ -653,144 +662,157 @@ If it is set to `Hash`, you will need to reference the `HASH_BY_INDEX` dictionar
...
+| 143  | InvalidTokenEventOrder            |
+| 144  | FailedToRecordMintEvent           |
+| 145  | FailedToRecordBurnedEvent         |
+| 146  | FailedToRecordApproveEvent        |
+| 147  | FailedToRecordTransferEvent       |
+| 148  | FailedToRecordMetadataUpdateEvent |
+| 149  | InvalidStartingEventId            |
+| 150  | InvalidLastEventId                |
+| 151  | MissingEventsMode                 |
+| 152  | InvalidEventsMode                 |
```
Comment (gRoussac) [inline README.md]: If my previous comment is correct 
```suggestion
| 141  | MissingEventsMode                 |
| 142  | InvalidEventsMode                 |
```

File: contract/src/modalities.rs [hunk 3]
```diff
@@ -210,6 +210,15 @@ impl TokenIdentifier {
     }
 }
 
+impl ToString for TokenIdentifier {
+    fn to_string(&self) -> String {
+        match self {
+            TokenIdentifier::Index(i) => i.to_string(),
+            TokenIdentifier::Hash(h) => h.to_string(),
```
Comment (gRoussac) [inline contract/src/modalities.rs]: Please avoid naming vars with one car

```suggestion
            TokenIdentifier::Index(index) => index.to_string(),
            TokenIdentifier::Hash(hash) => hash.to_string(),
```

File: tests/src/utility/constants.rs [hunk 4]
```diff
@@ -63,9 +63,10 @@ pub(crate) const ARG_IS_HASH_IDENTIFIER_MODE: &str = "is_hash_identifier_mode";
 pub(crate) const ARG_NAMED_KEY_CONVENTION: &str = "named_key_convention";
 pub(crate) const ARG_ACCESS_KEY_NAME_1_0_0: &str = "access_key_name";
 pub(crate) const ARG_HASH_KEY_NAME_1_0_0: &str = "hash_key_name";
+pub(crate) const ARG_EVENTS_MODE: &str = "events_mode";
 pub(crate) const ACCOUNT_USER_1: [u8; 32] = [1u8; 32];
 pub(crate) const ACCOUNT_USER_2: [u8; 32] = [2u8; 32];
-pub(crate) const ACCOUNT_USER_3: [u8; 32] = [2u8; 32];
```
Comment (gRoussac) [inline tests/src/utility/constants.rs:68]: I would expect this change to trigger some tests, strangely it does not seem to. 🤷

**Review by gRoussac — APPROVED** [review]
Please just check it does not add unnecessary errors to the already long list.

File: contract/src/utils.rs [hunk 5]
```diff
@@ -625,8 +625,7 @@ pub(crate) fn migrate_token_hashes(token_owner: Key) {
     storage::write(unmatched_hash_count_uref, unmatched_hash_count);
 }
 
-// This function is incredibly gas expensive
-// DO not use this function unless absolutely necessary.
+
```
Comment (gRoussac) [inline contract/src/utils.rs]: This additional line does not pass linting

File: contract/src/error.rs (hunk 1 above)
Comment (darthsiroftardis) [reply contract/src/error.rs]: Removed unused errors, I expect they were introduced in cep78 and have now become unnecessary

File: README.md (hunk 2 above)
Comment (darthsiroftardis) [reply README.md]: Updated README accordingly

File: contract/src/modalities.rs (hunk 3 above)
Comment (darthsiroftardis) [reply contract/src/modalities.rs]: Applied suggestion

File: contract/src/utils.rs (hunk 5 above)
Comment (darthsiroftardis) [reply contract/src/utils.rs]: Ran make lint

**Pull request merged**